stdlib/list.mli

```diff
@@ -52,16 +52,36 @@ val hd : 'a list -> 'a
 (** Return the first element of the given list. Raise
    [Failure "hd"] if the list is empty. *)
 
+val hd_opt: 'a list -> 'a option
+(** Return the first element of the given list or [None]
+    if the list is empty.
+    @since 4.05
+*)
+
 val tl : 'a list -> 'a list
 (** Return the given list without its first element. Raise
-   [Failure "tl"] if the list is empty. *)
+    [Failure "tl"] if the list is empty. *)
+
+val tl_opt: 'a list -> 'a list option
```

Comment:
I'm more strongly convinced that `hd_opt` and `tl_opt` are a bad idea than you were! `hd` and `tl` should only be used in the "get" scenario (i.e. where you know that the list is non-empty), otherwise you should be pattern matching the list (always), surely? Having `hd_opt` and `tl_opt` would seem to encourage poor constructs like:
```ocaml
match List.hd_opt l with
  Some hd -> ...
| None -> ...
```
or even worse:
```ocaml
List.hd_opt l |> Option.get |> ...
```
rather than the obvious:
```ocaml
match l with
  hd::_ -> ...
| [] -> ...
```
Perhaps a change to the docs for `hd` and `tl` explaining that philosophy, rather than these two functions? Or is there a really good scenario for wanting to use the `_opt` versions which I'm not seeing?